### ISSUE TITLE
Feature/function obj interface - [merged]

### ISSUE THREAD

**Issue by oliveroxtoby** [jaheyns/CfdOF#119]
In GitLab by @icojb25 on Apr 10, 2022, 09:14

_Merges feature/function_obj_interface -> develop_

First attempt at a function objects module, forces and forceCoeffs currently available, forces (pressure + viscous) plotted to XY-plot

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 10, 2022, 09:14

requested review from @oliveroxtoby

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 10, 2022, 09:18

added 1 commit

<ul><li>c0b0d6e9 - Changeup names to show forces, not monitor</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=370408980&start_sha=2353be3fbfdd564ec1b323d14e7647272c7fac3c)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 10, 2022, 09:19

added 14 commits

<ul><li>c0b0d6e9...402d05a3 - 13 commits from branch <code>develop</code></li><li>40bc1e37 - Merge branch &#39;develop&#39; into feature/function_obj_interface</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=370409061&start_sha=c0b0d6e906f688c86d5c1e8fcf2bd44a15d6dfeb)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 10, 2022, 09:22

marked this merge request as **ready**

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 11, 2022, 07:16

added 13 commits

<ul><li>40bc1e37...0ac50a4b - 12 commits from branch <code>develop</code></li><li>c4621212 - Merge branch &#39;develop&#39; into feature/function_obj_interface</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=370629373&start_sha=40bc1e37e841803bfbbe55247c21051828cb8000)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 12, 2022, 05:44

added 1 commit

<ul><li>e3250262 - Added some basic plotting for the force coefficients</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=371524133&start_sha=c4621212747dfef399aa0ae9926fcf3d05bcf397)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 12, 2022, 09:00

added 10 commits

<ul><li>e3250262...a2c681c4 - 9 commits from branch <code>develop</code></li><li>6cb3c3cc - Merge branch &#39;develop&#39; into feature/function_obj_interface</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=371695847&start_sha=e325026271f182185c20ef1724be6f757b518a2d)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 12, 2022, 09:08

added 1 commit

<ul><li>96f36e7b - Need to update plotters to OO&#39;s latest residual plot API</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=371704999&start_sha=6cb3c3ccd4631cac289f2144136427368a26ca9a)

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:01

added 1 commit

<ul><li>4d121589 - Add suggested icon for reporting function</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372096750&start_sha=96f36e7ba9dd318554c36352a7b93626c6e5e4e6)

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/functionobjects/CfdFunctionObjects.py line 86](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR86)



This shouldn't be needed, unless we want some colour coding. Probably don't need to worry for now.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Looking great, thanks a lot for this. 

One general request: I don't think we should try to to handle all function objects with one interface. There are so many different ones which play all kinds of different roles in OpenFOAM, I think it might be nightmare to try to handle them all together in the same way.

In terms of logical grouping, what comes to mind first are things like forces, forceCoeffs, min/max/mean of field values, and probes. One could perhaps categorise these all as 'Reporting functions' or something similar.

So in summary, my request would be to rename this from an interface for function objects, to an interface for reporting functions. I have also pushed a proposed icon that I think could work if you agree.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/functionobjects/CfdFunctionObjects.py line 98](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-9b87c35fa4546815f079324d954a5f9cR98)



I would not expose the field names to the user, as it's predetermined how it should be set up for each analysis type.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/functionobjects/CfdFunctionObjects.py line 108](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-9b87c35fa4546815f079324d954a5f9cR108)



It might be nice to automatically set this according to presence/absence of a porous zone in the analysis.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/functionobjects/CfdFunctionObjects.py line 120](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR120)



I would go for App::PropertyPoint for this for brevity (see e.g. ExtrusionAxisPoint in CfdMeshRefinement).

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/functionobjects/CfdFunctionObjects.py line 134](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR134)



I would use App::PropertyVector for the directions

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/gui/TaskPanelCfdFunctionObjects.ui line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-d8d2ddd07208c748e8de59933185bca4R1)



Although I originally asked for indents of one space as you have here (to be the same as my qtdesigner), I think we concluded we should standardise on four spaces? Would that make sense?

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/plotters/CfdForceCoeffPlot.py line 45](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-83d2219e6d7b471fe5432243ac908622R45)



residuals_forces -> forces

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [core/plotters/CfdForcesPlot.py line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-c54a77039e623ad410f5ee8fb4de09f7R1)



I think it should be possible to use one class for both forces and forceCoeffs by passing in a few parameters. Might be nice to avoid too much code duplication.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [data/defaults/system/controlDict line 111](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..8e49881325d00c5b80245a1b5fe9ba51b433d0ca#diff-91110dc26b9e619684e4eaab56df78e5R111)



Uncomment?

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 12, 2022, 15:02

Commented on [CfdRunnableFoam.py line 200](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-61409626b02102a5d6a53be8931ae4ddR200)



Rename ...Residuals -> ...Forces

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:18

Commented on [core/plotters/CfdForcesPlot.py line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-c54a77039e623ad410f5ee8fb4de09f7R1)



hi @oliveroxtoby yes, of course. As indicated, I was pushing some code as a first impression. There is obviously a lot of refactoring that can happen. I've have just tried to get working POC's into the repo as a starting point. I fact, I think a single parameterised class for all the plotting we have at the moment will suffice.

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:20

Commented on [data/defaults/system/controlDict line 111](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..8e49881325d00c5b80245a1b5fe9ba51b433d0ca#diff-91110dc26b9e619684e4eaab56df78e5R111)



Not yet, there is some review on how we want to / how best / a standardised way to input vector data. The one input per component is tiresome to say the least. Again, I wasnt sure if we would even want this inclusion, in keeping with the whole "keeping it simple / minimal fiddle factors etc" ethos. LMK :grin:

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:21

Commented on [core/functionobjects/CfdFunctionObjects.py line 120](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR120)



Thanks, this is what I was looking for, unfortunately there is almost NIL on the internal API. Even the Doxygen stuff

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:28

Yes, that was the intention, hence only Forces and ForceCoeffs. There are a lot of interesting one I noticed (interesting as I've never really looked deeply at whats on offer). I think if we jot down our thoughts on the ROADMAP or somewhere else, it will might make minimise rework etc (esp if the "idea" comes from the other party). 

Icon looks cool. Only comment is that the others are PNG and its generally good to standardise formats (or have sets of formats depending on OS's etc). Most OS's I have worked on use PNG or ICO (which is layered PNG's). I also noted the icons are all different sizes (I tried to standard to options on what seemed to be the most common h x w for the existing ones) and we need to standardise the naming convention, I just didnt get time to do that. Only comment is whether we'd want to standardise on the colours (dark grey / green / while / blue) or include reds as well. Subjective I guess. I think I'd personally standardise on a single format, potentially PNG.

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:35

Commented on [core/gui/TaskPanelCfdFunctionObjects.ui line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-d8d2ddd07208c748e8de59933185bca4R1)



:thinking: OK, i might have gotten confused - I thought you mentioned your GUI builder was non-configurable which is why I changed to what you were (are) using, so I assumed it would give what we agreed. I have not found a preference to change the indent in Qt Creator (same as you, unless you know different?)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:40

Commented on [core/plotters/CfdForceCoeffPlot.py line 45](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-83d2219e6d7b471fe5432243ac908622R45)



As with the Porous zones (auto-enablement) etc, I've tried to break the back of the bulk of the functionality to start with rather than cover every nook and cranny and potential refinement - feel free to pull and make little changes like these if required. Happy for any to add to and / or refine / make better. So I guess more of a team relay / collaboration rather than individual submissions. In this instance, I've tried to do the bulk of the heavy lifting if that makes sense but obviously time is a constraint as well.

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 04:41

Commented on [core/functionobjects/CfdFunctionObjects.py line 86](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR86)



Agreed, will remove

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 05:34

Commented on [core/functionobjects/CfdFunctionObjects.py line 134](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR134)



thanks!

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 06:22

Commented on [core/functionobjects/CfdFunctionObjects.py line 86](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR86)



changed this line in [version 9 of the diff](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372557556&start_sha=4d1215896e2a86f4a81df385c9d0873005e7e606#d2655e4fe28ab39f8af9da01ccee54f2e2a4e6ec_86_86)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 06:22

Commented on [core/functionobjects/CfdFunctionObjects.py line 120](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR120)



changed this line in [version 9 of the diff](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372557556&start_sha=4d1215896e2a86f4a81df385c9d0873005e7e606#d2655e4fe28ab39f8af9da01ccee54f2e2a4e6ec_120_107)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 06:22

Commented on [core/functionobjects/CfdFunctionObjects.py line 134](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..4d1215896e2a86f4a81df385c9d0873005e7e606#diff-9b87c35fa4546815f079324d954a5f9cR134)



changed this line in [version 9 of the diff](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372557556&start_sha=4d1215896e2a86f4a81df385c9d0873005e7e606#d2655e4fe28ab39f8af9da01ccee54f2e2a4e6ec_134_115)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 06:22

added 1 commit

<ul><li>8e498813 - Use FC vectors / points API, parameterise residuals plot, update UI</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372557556&start_sha=4d1215896e2a86f4a81df385c9d0873005e7e606)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 10:06

Commented on [data/defaults/system/controlDict line 111](https://github.com/jaheyns/CfdOF/compare/a2c681c4629161a373d1d549deacb2fd570333f6..8e49881325d00c5b80245a1b5fe9ba51b433d0ca#diff-91110dc26b9e619684e4eaab56df78e5R111)



changed this line in [version 10 of the diff](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372773421&start_sha=8e49881325d00c5b80245a1b5fe9ba51b433d0ca#d7f1336abe5a2a16a112dca157cbf1249fed8cdd_111_109)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 10:06

added 1 commit

<ul><li>c2dc9a4b - Tidy controlDict, test, preprocess function objects to strip out FC nomenclature</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372773421&start_sha=8e49881325d00c5b80245a1b5fe9ba51b433d0ca)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 10:14

added 4 commits

<ul><li>c2dc9a4b...25e49f11 - 3 commits from branch <code>develop</code></li><li>105673f8 - Merge branch &#39;develop&#39; into feature/function_obj_interface</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=372781713&start_sha=c2dc9a4b6c1c08d33438b8ffac913896a3bed53b)

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 13:55

added 1 commit

<ul><li>def2a541 - Continue integrating new residuals plot into forces / coeffs plots</li></ul>

[Compare with previous version](/opensimproject/cfdof/-/merge_requests/21/diffs?diff_id=373010577&start_sha=105673f8f286b0bef5547f284ba678396f76fbba)

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 13, 2022, 14:04

Cool. To clarify, my comment was mostly just directed at the naming - I would use CfdReportingFunctions rather than CfdFunctionObjects, as I think you'll need a different interface for other types of function objects.

We started with png icons but as FreeCAD supports svg, which is infinitely/losslessly scalable, this seems like the way to go and I am in the process of transitioning over. It means re-drawing the old icons though, so not will take a while.

Yup, I wondered about the colour scheme as well. We can try a few things and see what looks best.

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 13, 2022, 14:11

Commented on [core/gui/TaskPanelCfdFunctionObjects.ui line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-d8d2ddd07208c748e8de59933185bca4R1)



That's right, I was asking for one space initially but as you said you preferred larger indents I was offering to reformat to four spaces (https://gitlab.com/opensimproject/cfdof/-/merge_requests/10#note_902014601). But if you're fine with tolerating the single space indents then we don't have to worry about reformatting so all good with me...

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 13, 2022, 14:15

approved this merge request

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 13, 2022, 14:25

Thanks for another great contribution. Sorry if I offended - I only meant to offer my perspective. 

Perhaps it's better if you work directly in develop and pop me a quesition if you want my input on something.

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 14:26

OK, got it. I'll change the name :thumbsup: 

Re: SVG - ok cool. Should we standardise a (default) template size (maybe 512 x 512)?

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 14:28

Commented on [core/gui/TaskPanelCfdFunctionObjects.ui line 1](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-d8d2ddd07208c748e8de59933185bca4R1)



OK cool - yeah, i did (do) but it doesnt seem like you can set this in Creator. Now that I am using Creator (not editing the XML), is not really that important from my side.

**Comment by oliveroxtoby** [comment]
In GitLab by @icojb25 on Apr 13, 2022, 14:30

Commented on [CfdRunnableFoam.py line 200](https://github.com/jaheyns/CfdOF/compare/25e49f116956ea76ee76991423bc1e16aba5469a..def2a541c04c459521e702dc2782aeaa6f85a9bb#diff-61409626b02102a5d6a53be8931ae4ddR200)



I've removed the additional plotters, they all use the original Residual Plot class now? LMK you're good with that. Unless we're going to do something very different (like subplots) I think this should be fine?

**Comment by oliveroxtoby** [comment]
In GitLab by @oliveroxtoby on Apr 14, 2022, 13:44

Sure, sounds good.
